src/Enable.Extensions.Queuing.RabbitMQ/Internal/ConnectionManagement/RabbitMQConnectionManager.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Enable.Extensions.Queuing.RabbitMQ.Internal.ConnectionManagement;
+using RabbitMQ.Client;
+using RabbitMQ.Client.Exceptions;
+
+namespace Enable.Extensions.Queuing.RabbitMQ.Internal
+{
+    internal class RabbitMQConnectionManager
+    {
+        private static readonly List<(int hash, RabbitMQManagedConnection managedConnection)>
```

Comment:
A list of an anonymous tuple doesn't feel like the right type to be using here. Would a dictionary be more appropriate? Furthermore, there are a lot of manual locks used throughout this class. These could be removed if we used a `ConcurrentDictionary` here.

Reply:
This List of type tuple is being used as the list contains tuples with duplicate keys. This is so that different instances of the connection manager can find a connection with channels available, based on the the hash of the connection options.

Manual locks are used to ensure multiple threads do not try to create multiple connections at once if they don't find one in the existing connections.